src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs

```diff
@@ -96,57 +109,64 @@ public async Task<IEnumerable<NuGetReference>> GetNewerVersionsAsync(NuGetRefere
                 }
             }
 
-            // Filter to only include versions higher than the user's current version and,
-            // optionally, only the highest minor/build for each major version
-            var currentVersion = reference.GetNuGetVersion();
-            var filteredVersions = versions.Distinct().Where(v => v > currentVersion);
-            var versionsToReturn = latestMinorAndBuildOnly
-                ? filteredVersions.GroupBy(v => v.Major).Select(v => v.Where(v => !v.IsPrerelease).Max() ?? v.Max()!)
-                : filteredVersions;
+            var tfmSet = ImmutableHashSet.CreateRange(tfms.Select(t => NuGetFramework.Parse(t.Name)));
 
-            _logger.LogDebug("Found versions for package {PackageName}: {PackageVersions}", reference.Name, versionsToReturn);
-
-            return versionsToReturn.OrderBy(v => v).Select(v => reference with { Version = v.ToNormalizedString() });
+            return FilterSearchResults(name, results, tfmSet, currentVersion, latestMinorAndBuildOnly);
         }
 
-        public async Task<NuGetReference?> GetLatestVersionAsync(string packageName, bool includePreRelease, string[]? packageSources, CancellationToken token)
+        private static IEnumerable<NuGetReference> FilterSearchResults(string name, List<IPackageSearchMetadata> searchResults, ImmutableHashSet<NuGetFramework> tfms, NuGetVersion? currentVersion = null, bool latestMinorAndBuildOnly = false)
         {
-            NuGetVersion? highestVersion = null;
+            if (searchResults.Count == 0)
+            {
+                return Enumerable.Empty<NuGetReference>();
+            }
+
+            var results = searchResults
+                .Where(r => currentVersion is null ? true : r.Identity.Version > currentVersion);
 
-            // Query each package source for listed versions of the given package name
-            foreach (var source in packageSources?.Select(p => new PackageSource(p)) ?? _packageSources)
+            if (latestMinorAndBuildOnly)
             {
-                try
-                {
-                    var metadata = await GetSourceRepository(source).GetResourceAsync<PackageMetadataResource>(token).ConfigureAwait(false);
-                    var searchResults = await CallWithRetryAsync(() => metadata.GetMetadataAsync(packageName, includePrerelease: includePreRelease, includeUnlisted: false, _cache, _nugetLogger, token)).ConfigureAwait(false);
-                    var highestVersionResult = searchResults.Select(r => r.Identity.Version).Max(v => v);
-                    if (highestVersionResult is null)
-                    {
-                        continue;
-                    }
+                var max = results
+                    .GroupBy(r => r.Identity.Version)
```

Comment:
Shouldn't we be grouping by Version.Major for the `latestMinorAndBuildOnly` path?